src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -226,17 +226,23 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
                 }
             }
 
-            while (tasks?.Count > 0)
+            if (tasks?.Count > 0)
             {
-                // Wait for any node to finish resolving
-                var task = await Task.WhenAny(tasks);
-
-                // Extract the resolved node
-                tasks.Remove(task);
-                var dependencyNode = await task;
-                dependencyNode.OuterNode = node;
+                if (tasks.Count > 1)
+                {
+                    // Wait for all the nodes to finish resolving. We only do this if we
+                    // have more than one to avoid WhenAll's defensive copying allocations.
+                    // Otherwise, if there's just one, we'll just end up waiting on it in
+                    // the loop.
+                    await Task.WhenAll(tasks);
```

Comment:
Is there a benefit to doing `await Task.WhenAll(tasks)`, rather than just letting the `await task` in the `foreach` loop below effectively handle it?

Is it exception handling to return all exceptions in an `AggregateException`, rather than bailing out at the first exception, like our old code did?

Reply:
First one, yes, this would result in the method resuming on the thread pool for every dependency, only to be immediately blocked on the next one. This results in a single callback when they are all done.
Second one, I don't think exception handling is any different here, can you expand?

Reply:
From [Task.WhenAll's docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?view=net-5.0)

> If any of the supplied tasks completes in a faulted state, the returned task will also complete in a Faulted state, where its exceptions will contain the aggregation of the set of unwrapped exceptions from each of the supplied tasks.

In other words, imagine a node (package) in the graph has 10 dependencies, and walking the graph for 2 of them result in unrecoverable HTTP errors. In the old code, and if we did not do the `Task.WhenAll`, `await task` will bubble up the `NuGet.Protocol.ProtocolException` that wraps the `HttpRequestException`, and this "single" exception will bubble up the call stack. The second faulted task will never be handled. With `await Task.WhenAll`, both faulted tasks `ProtocolExceptions` will be put in an `AggregateException`, and that will bubble up.

I didn't realise it when I wrote my original question, but this is might actually change the contract of the API, since the exception type might be changing. This might not be a problem, because NuGet doesn't have particularly good error handling and error messages. But I think we should make a small effort to check if `ProtocolException` is caught higher up in the call stack, and if so, consider improving it to handle `AggregateException` in addition.

However, it might not be necessary, because my memory is that async methods created by the compiler always throw `AggregateException`, even if there's only a single exception, and that `await faultedTask` will always re-throw `aggregateException.InnerExceptions[0]`, ignoring any others.

Async is complex 😕  I think I've convinced myself that this change is fine-as is, and is not breaking, unlike what I wrote two paragraphs up.

Reply:
> However, it might not be necessary, because my memory is that async methods created by the compiler always throw AggregateException, even if there's only a single exception, and that await faultedTask will always re-throw aggregateException.InnerExceptions[0], ignoring any others.

I think that is correct, the headache of handling  `AggregateExceptions` is over after the introduction of `await` keyword. 

BTW, `await Task.WhenAll(tasks)` is necessary to avoid problem with [unobserved exceptions](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler.unobservedtaskexception?view=net-5.0). 

For example, code below is broken:
```
    var t1 = Foo();
    var t2 = Bar();
    await t1;
    await t2;
```    
If `t1` throws exception then `t2` is never going to be awaited. When `t2` will be collected by GC and it will be in faulted state it will cause the `UnobservedTaskException` being thrown.